Added workflow for rollback release

kumarnarendra701 opened the pull request:
<!--
Added rollback workflow for canary, virtualenv, and prod.
-->

kumarnarendra701 committed:
Added manual workflow for rollback

kumarnarendra701 committed:
updated final fix

cpswan commented on .github/workflows/at_server.yaml:
```diff
@@ -515,7 +530,7 @@ jobs:
     # Runs only after functional tests are completed.
     needs: [ unit_tests, functional_tests, end2end_test_12, end2end_test_34, end2end_test_56 ]
-    if: ${{ github.repository == 'atsign-foundation/at_server' && github.event_name == 'push' && contains(github.ref, 'refs/tags/c') }}
+    if: ${{ github.repository == 'atsign-foundation/at_server' && github.event_name == 'push' && contains(github.ref, 'refs/tags/c') || github.event.inputs.rollback_canary_secondary_image == 'true' }}
```
In the steps below it shouldn't be building a new Docker image. What needs to happen instead is that the appropriate tag gets pointed back to the selected previous version.

So let's say that my latest release is `v3.0.18` and I want to roll back to `v3.0.17` the script needs to create and push manifest for the `prod` tag that's the same as the `prod-v3.0.17` manifest.

kumarnarendra701 committed:
Added rollback steps for canary and prod secaondary

cpswan commented on .github/workflows/at_server.yaml:
```diff
@@ -685,6 +698,83 @@ jobs:
+
+      # Extract version for docker tag
+      - name: Get version
+        run: echo "VERSION=${GITHUB_REF##*/}" >> $GITHUB_ENV
```
How does this get a correct version for the prod tag we want to roll back to?

kumarnarendra701 replied:
Fixed

cpswan commented on .github/workflows/at_server.yaml:
```diff
@@ -685,6 +698,83 @@ jobs:
+      - name: Get version
+        run: echo "VERSION=${GITHUB_REF##*/}" >> $GITHUB_ENV
+
+      - name: Set up QEMU
```
Why do we need QEMU?

kumarnarendra701 replied:
Removed this in the final fix.

cpswan commented on .github/workflows/at_server.yaml:
```diff
@@ -685,6 +698,83 @@ jobs:
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2.0.0
+
+      - name: Set up Docker Buildx
```
Why do we need BuildX?

kumarnarendra701 replied:
Removed this in the final fix.

cpswan commented on .github/workflows/at_server.yaml:
```diff
@@ -685,6 +698,83 @@ jobs:
+        run: |
+          docker pull --platform=linux/amd64 atsigncompany/secondary:prod-${{ env.VERSION }}
+          docker tag atsigncompany/secondary:prod-${{ env.VERSION }} atsigncompany/secondary:prod-${{ env.VERSION }}-amd64
+          docker push atsigncompany/secondary:prod-${{ env.VERSION }}-amd64
```
Do we need to push, or can the manifest be created locally?

kumarnarendra701 replied:
Fixed

cpswan commented on .github/workflows/at_server.yaml:
```diff
@@ -685,6 +698,83 @@ jobs:
+        run: |
+          docker pull --platform=linux/arm64 atsigncompany/secondary:prod-${{ env.VERSION }}
+          docker tag atsigncompany/secondary:prod-${{ env.VERSION }} atsigncompany/secondary:prod-${{ env.VERSION }}-arm64
+          docker push atsigncompany/secondary:prod-${{ env.VERSION }}-arm64
```
Again do we need to push?

kumarnarendra701 replied:
Fixed

cpswan commented on .github/workflows/at_server.yaml:
```diff
@@ -685,6 +698,83 @@ jobs:
+            --amend atsigncompany/secondary:prod-${{ env.VERSION }}-amd64
+          docker manifest push atsigncompany/secondary:prod
+
+  # Rollback canary secondary image.
```
Same comments apply here as for prod

kumarnarendra701 replied:
Fixed

cpswan reviewed (changes requested):
@kumarnarendra701 I'm not yet confident that this will work as expected.

Can you maybe set up a test repo (using test images) to prove the plumbing.

I've previously used https://github.com/cpswan/dockerfiles/tree/main/hello_ver to create images that do nothing much more than carry different versions around for the purpose of testing.

cpswan committed:
Merge branch 'trunk' into feature/rollback-release

kumarnarendra701 commented:
@cpswan - I have created a test repo with rollback flow. Please verify it and let me know if you see any enhancement.
https://github.com/atsign-foundation/devops-workflow-test

For reference.
![image](https://user-images.githubusercontent.com/29299536/178114243-35d153a0-1675-4fd6-aa9c-a4ccac3b0684.png)

cpswan commented:
Thanks @kumarnarendra701 some really good progress here.

Looking at your test repo I see different digests for prod, v3.0.17, v3.0.16 and v3.0.15

What we want to be able to do here is choose which version prod gets reverted to, and then see the same digest for (say) v3.0.16 and prod. And to be able to do the same thing for canary.

kumarnarendra701 commented:
@cpswan - It's creating a different digest after pulling ':${{ env.VERSION }}' image creates a new tag and pushes it to the docker hub and then it creates a new digest for amd64 and arm64 the same digest available in the manifest.

I have reverted it to v3.0.16 and created two tags for amd64 and arm64 and also manifest has the same digest.

![image](https://user-images.githubusercontent.com/29299536/178696464-fc73b5fe-7dec-4735-a6b0-7f142110eb22.png)

How I test it - 
1. Pulled the latest ':prod' image that was pointing to [v3.0.17]
```
narendrakumar@Narendras-MacBook-Pro-2 ~ % docker pull kumarnarendra701/docker-rollback:prod
prod: Pulling from kumarnarendra701/docker-rollback
1a07b0163994: Pull complete 
da8dc61ccaca: Pull complete 
59be51dca150: Pull complete 
1f391fc42cee: Pull complete 
624799215cc8: Pull complete 
10e7e30b294e: Pull complete 
1845118be0cd: Pull complete 
c8e041c4002a: Pull complete 
c0b7948555b6: Pull complete 
73a6b1154f15: Pull complete 
12e8e47fb13c: Pull complete 
c2e806cdeee5: Pull complete 
78884ba77c45: Pull complete 
7a48a5d9e933: Pull complete 
Digest: sha256:2e6b99028568a37fea47f92886d75890203b8a2e4c211d74dbbf80f8e260a09e
Status: Downloaded newer image for kumarnarendra701/docker-rollback:prod

narendrakumar@Narendras-MacBook-Pro-2 ~ % docker run -it kumarnarendra701/docker-rollback:prod bash 
root@5f9692686efa:/var/www/html# ls
test.txt
root@5f9692686efa:/var/www/html# cat test.txt 
**hello world v3.0.17**
```

2. Reverted it to [v3.0.16] and then pulled the latest image and verify the content.
```
 docker pull kumarnarendra701/docker-rollback:prod           
prod: Pulling from kumarnarendra701/docker-rollback
1a07b0163994: Already exists 
da8dc61ccaca: Already exists 
59be51dca150: Already exists 
1f391fc42cee: Already exists 
624799215cc8: Already exists 
10e7e30b294e: Already exists 
1845118be0cd: Already exists 
c8e041c4002a: Already exists 
c0b7948555b6: Already exists 
73a6b1154f15: Already exists 
12e8e47fb13c: Already exists 
c2e806cdeee5: Already exists 
78884ba77c45: Already exists 
ee1e274bbb6d: Pull complete 
Digest: sha256:9d473d348625730c7c4ddfe6c5a59f6a05c68e94b5a5de04f20d6bf8f7fd64ea
Status: Downloaded newer image for kumarnarendra701/docker-rollback:prod
docker.io/kumarnarendra701/docker-rollback:prod
narendrakumar@Narendras-MacBook-Pro-2 ~ % docker run -it kumarnarendra701/docker-rollback:prod bash
root@c717c38a2a56:/var/www/html# ls
test.txt
root@c717c38a2a56:/var/www/html# cat test.txt 
**hello world v3.0.16**
```
Note - After reverting the build, the content of test.txt changes to the previous version.

Reg:  And to be able to do the same thing for the canary. - Yes, I'll do the same for the canary also. 
Are you looking for something another step "Deploy Canary to Prod" it means to pull the canary image and push it to the secondary image for the #750 ticket.

```
docker pull --platform=linux/amd64  atsigncompany/secondary:canary-${{ env.VERSION }}
docker tag atsigncompany/secondary:canary-${{ env.VERSION }}atsigncompany/secondary:prod-${{ env.VERSION }}-amd64
docker push atsigncompany/secondary:prod-${{ env.VERSION }}-amd64
```

cpswan commented:
@kumarnarendra701 I'm still not clear how the version to be rolled back to is selected?

Yes, we do need the ability to promote canary to prod, which is what #750 is for. But note that we do need to create a new Docker image in that case. Where we start `FROM atsigncompany/secondary:canary` then copy in a new pubspec.yaml with the correct version number in it so that it has a prod version number rather than a canary version number.

kumarnarendra701 committed:
Update at_server.yaml

kumarnarendra701 committed:
Update at_server.yaml

kumarnarendra701 committed:
Update at_server.yaml

cpswan reviewed (commented):
Is this still needed, or overtaken by #817 ?

kumarnarendra701 commented:
@cpswan - We have the latest code(rollback + deploy canary to prod) on https://github.com/atsign-foundation/at_server/pull/817. If you approve this PR then we can ignore this PR(#779).

cpswan commented:
Thanks @kumarnarendra701 closing this so we can concentrate on #817, which I have approved (though there is a lingering question).
